### PR TITLE
Fixes for #2337, #2349, #2288, #2333, #1398.

### PR DESCRIPTION
**Pull request by subtly**

### CODE REVIEW

File: libp2p/Host.cpp [hunk 1]
```diff
@@ -170,15 +184,13 @@ void Host::doneWorking()
 	// disconnect pending handshake, before peers, as a handshake may create a peer
 	for (unsigned n = 0;; n = 0)
 	{
-		{
-			Guard l(x_connecting);
+		DEV_GUARDED(x_connecting)
 			for (auto i: m_connecting)
```
Comment (chfast) [inline libp2p/Host.cpp]: You should considered iterating over the shared/weak pointers using the reference, not the copy. I.e. `for (auto& i: m_connecting)`. For performance reasons you may not want to temporary bump the shared/weak counter.

There is also "alwyas correct" variant of the for loop: `for (auto&& x: collection)`. That one will be preferred by later C++ standards by adding the syntax sugar in the form of `for(x: collection)`.

**Comment by gavofyork** [comment]
compile fail on osx?

File: libdevcore/Worker.cpp [hunk 2]
```diff
@@ -50,11 +50,16 @@ void Worker::startWorking()
...
-				workLoop();
-//				cnote << "Finishing up worker thread...";
-				doneWorking();
+				try
+				{
+					startedWorking();
+					workLoop();
+					doneWorking();
+				}
+				catch (std::exception const& _e)
```
Comment (arkpar) [inline libdevcore/Worker.cpp]: what about boost exceptions? Note that boost::current_exception_diagnostic_information() works for both

File: libp2p/Host.cpp [hunk 3]
```diff
@@ -120,6 +120,15 @@ Host::~Host()
 void Host::start()
 {
 	startWorking();
+	while (isWorking() && !haveNetwork())
+		this_thread::sleep_for(chrono::milliseconds(10));
```
Comment (gluk256) [inline libp2p/Host.cpp]: i suggest a timeout here, instead of eternal cycle.

**Comment by gavofyork** [comment]
tests failing.

File: libdevcore/Worker.cpp (hunk 2 above)
Comment (subtly) [inline libdevcore/Worker.cpp]: Happy to change this to anything more useful.

File: libp2p/Host.cpp (hunk 3 above)
Comment (subtly) [inline libp2p/Host.cpp]: isWorking will be set to false if/when startWorking fails due to an exception; if the exception doesn't occur then haveNetwork will return true.

File: libdevcore/Worker.cpp (hunk 2 above)
Comment (arkpar) [inline libdevcore/Worker.cpp]: I mean something like this:
```
catch(...)
{
    clog(WarnChannel) << boost::current_exception_diagnostic_information();
}
```

**Pull request merged**